### PR TITLE
[WIP]further meta integration

### PR DESCRIPTION
**Pull request by ghost**
add watch from beginning to tvshows and movies if guide has data for it and meta is installed.
it emulates delayed viewing through meta
 it requires a datase reset to start working.

### CODE REVIEW

**Comment by ghost** [comment]
depends on https://github.com/metate/meta4kodi/pull/31

File: script.ftvguide/source.py [hunk 1]
```diff
@@ -967,8 +985,34 @@ def parseXMLTV(self, context, f, size, logoFolder, progress_callback):
...
+                        categories = elem.findall("category")
+                        for category in categories:
+                            if category.text == "Movies" or channel.lower().find("sky movies") != -1:
+                                is_movie = "Movie"
+                                break
+
+                        if episode_num is not None:
+                                if str.find(episode_num, ".") != -1:
+                                    splitted = str.split(episode_num, ".")
+                                    season = int(splitted[0]) + 1
```
Comment (bluezed) [inline script.ftvguide/source.py:1015]: @midraal 
Getting an exception here:

```
07:47:24 T:9560   DEBUG: Settings changed: True
07:47:24 T:9560   DEBUG: Database.eventLoop() >>>>>>>>>> processing command: _getEPGView
07:47:24 T:9560   DEBUG: [script.ftvguide] Updating caches...
07:47:24 T:9560    INFO: XCURL::DllLibCurlGlobal::easy_aquire - Created session to ftp://192.168.2.1
07:47:30 T:9560   ERROR: Traceback (most recent call last):
07:47:30 T:9560   ERROR:   File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\script.ftvguide\source.py", line 365, in _updateChannelAndProgramListCaches
07:47:30 T:9560   ERROR:     for item in self.source.getDataFromExternal(date, progress_callback):
07:47:30 T:9560   ERROR:   File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\script.ftvguide\source.py", line 1003, in parseXMLTV
07:47:30 T:9560   ERROR:     season = int(splitted[0]) + 1
07:47:30 T:9560   ERROR: ValueError: invalid literal for int() with base 10: ''
```

**Comment by bluezed** [comment]
Thanks for the work you've done!
Just tested it here and it fails at the position I marked in the diff...

Also, can you find a way to gracefully migrate the database changes you need rather than forcing a full reset? If a user doesn't reset the guide just gets stuck in an endless loading loop...

Thank you!

**Comment by ghost** [comment]
i think this fixed it, might need some more testing.
also needs a guide refresh after migration

**Comment by bluezed** [comment]
Thanks, the change seems to work ok now :+1: 

Couple more things now...
- Getting this error in the log when I try to use "Watch from beginning", only for some shows though...:

```
09:10:14 T:10588   DEBUG: [xbmcswift2] Handling incoming request for /tv/play_by_name/MacGyver/3/8
09:10:14 T:10588  NOTICE: [xbmcswift2] Request for "/tv/play_by_name/MacGyver/3/8" matches rule for function "tv_play_by_name"
09:10:15 T:10588   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <type 'exceptions.KeyError'>
                                            Error Contents: ('firstaired',)
                                            Traceback (most recent call last):
                                              File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\plugin.video.meta\addon.py", line 177, in <module>
                                                main()
                                              File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\plugin.video.meta\addon.py", line 174, in main
                                                plugin.run()
                                              File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\plugin.video.meta\resources\lib\xbmcswift2\plugin.py", line 333, in run
                                                items = self._dispatch(self.request.path)
                                              File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\plugin.video.meta\resources\lib\xbmcswift2\plugin.py", line 306, in _dispatch
                                                listitems = view_func(**items)
                                              File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\plugin.video.meta\resources\lib\meta\navigation\tvshows.py", line 105, in tv_play_by_name
                                                show["year"] = int(show['firstaired'].split("-")[0].strip())
                                            KeyError: ('firstaired',)
                                            -->End of Python script error report<--
```
- Other shows don't bring an error but nothing seems to happen. Here's what's in the log:

```
09:18:29 T:10996   DEBUG: [xbmcswift2] Handling incoming request for /tv/play_by_name/The%20Big%20Bang%20Theory/7/21
09:18:29 T:10996  NOTICE: [xbmcswift2] Request for "/tv/play_by_name/The%20Big%20Bang%20Theory/7/21" matches rule for function "tv_play_by_name"
09:18:29 T:10996 WARNING: [xbmcswift2] No converter provided, unicode should be used, but returning str value
09:18:29 T:10996 WARNING: Previous line repeats 1 times.
09:18:29 T:10996   ERROR: XFILE::CDirectory::GetDirectory - Error getting C:\Users\Thomas\AppData\Roaming\Kodi\userdata\addon_data\plugin.video.meta\players\
09:18:29 T:10996   ERROR: XFILE::CDirectory::GetDirectory - Error getting special://profile/addon_data/plugin.video.meta/players/
09:18:29 T:10996 WARNING: Attempt to use invalid handle -1
09:18:29 T:10996    INFO: CPythonInvoker(29, C:\Users\Thomas\AppData\Roaming\Kodi\addons\plugin.video.meta\addon.py): script successfully run
```

Guess maybe we need an indication that nothing was found or similar?
- Also, the "Watch from beginning" does not appear for any Movies. Tried "Red" and "Sudden Impact" that were on E4 and Dave last night...
- Finally, could you add the new dialog option to all the skins please?

Overall, if this all works then it'd make a great addition to the guide :grinning:

**Comment by ghost** [comment]
can you try the latest commit to  metate/meta4kodi#31 ? it should fix the tvshow stuff.
not all movies or shows are going to work unfortunately, it depends on them being labeled as such in guide.xmltv.
if it shows (movie) or S{season}E{episode} at the end of the info title it should work tough

**Comment by bluezed** [comment]
Ah ok about the Movies...

The earlier error seems to be sorted but getting a new one now:

```
10:36:24 T:7880   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <type 'exceptions.UnicodeEncodeError'>
                                            Error Contents: 'ascii' codec can't encode character u'\u2013' in position 73: ordinal not in range(128)
                                            Traceback (most recent call last):
                                              File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\script.ftvguide\gui.py", line 214, in onAction
                                                self.onActionEPGMode(action)
                                              File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\script.ftvguide\gui.py", line 337, in onActionEPGMode
                                                self._showContextMenu(program)
                                              File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\script.ftvguide\gui.py", line 436, in _showContextMenu
                                                program.title.replace(" ", "%20").replace(",",""), program.season, program.episode))
                                            UnicodeEncodeError: 'ascii' codec can't encode character u'\u2013' in position 73: ordinal not in range(128)
                                            -->End of Python script error report<--
```

That's when I select this show (from a German TV channel):
Criminal Intent - Verbrechen im Visier S1E9

**Comment by ghost** [comment]
looks like it's having issues with unicode, i'll have a look

**Comment by ghost** [comment]
this should hopefully fix that bug, but i couldn't find any programs with such a dash in my guide so i can't test it

**Comment by bluezed** [comment]
Sorry, still same issue:

```
11:21:58 T:11956   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <type 'exceptions.UnicodeEncodeError'>
                                            Error Contents: 'ascii' codec can't encode character u'\u2013' in position 73: ordinal not in range(128)
                                            Traceback (most recent call last):
                                              File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\script.ftvguide\gui.py", line 214, in onAction
                                                self.onActionEPGMode(action)
                                              File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\script.ftvguide\gui.py", line 337, in onActionEPGMode
                                                self._showContextMenu(program)
                                              File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\script.ftvguide\gui.py", line 437, in _showContextMenu
                                                title, program.season, program.episode))
                                            UnicodeEncodeError: 'ascii' codec can't encode character u'\u2013' in position 73: ordinal not in range(128)
                                            -->End of Python script error report<--
```

Here's my XMLTV file... the channel is called "VOX" and it's got some more programmes with the same dash. [my_guide.xmltv](https://filetea.me/t1sJeidtMmnRySP3LQCiIxfjA)

**Comment by ghost** [comment]
didn't add a migration from last commit to this one so needs a database reset. also needs latest commit in my meta branch. it should now also be able to find non-english shows/movies

**Comment by bluezed** [comment]
Ok... seems to almost work now :wink: 

Getting this error now with some shows:

```
14:58:11 T:9164   DEBUG: [xbmcswift2] Handling incoming request for /tv/play_by_name/MacGyver/3/9/de
14:58:11 T:9164  NOTICE: [xbmcswift2] Request for "/tv/play_by_name/MacGyver/3/9/de" matches rule for function "tv_play_by_name"
14:58:12 T:9164   ERROR: [plugin.video.meta] missing translation for choose show
14:58:12 T:9164   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <type 'exceptions.TypeError'>
                                            Error Contents: descriptor 'encode' requires a 'unicode' object but received a 'str'
                                            Traceback (most recent call last):
                                              File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\plugin.video.meta\addon.py", line 177, in <module>
                                                main()
                                              File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\plugin.video.meta\addon.py", line 174, in main
                                                plugin.run()
                                              File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\plugin.video.meta\resources\lib\xbmcswift2\plugin.py", line 333, in run
                                                items = self._dispatch(self.request.path)
                                              File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\plugin.video.meta\resources\lib\xbmcswift2\plugin.py", line 306, in _dispatch
                                                listitems = view_func(**items)
                                              File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\plugin.video.meta\resources\lib\meta\navigation\tvshows.py", line 120, in tv_play_by_name
                                                unicode.encode(s["seriesname"], "ascii", "ignore"), s["year"]) for s in items])
                                            TypeError: descriptor 'encode' requires a 'unicode' object but received a 'str'
                                            -->End of Python script error report<--
```

```
15:00:13 T:15576   DEBUG: [xbmcswift2] Handling incoming request for /tv/play_by_name/Baby%20Daddy/4/20/en
15:00:13 T:15576  NOTICE: [xbmcswift2] Request for "/tv/play_by_name/Baby%20Daddy/4/20/en" matches rule for function "tv_play_by_name"
15:00:14 T:15576   ERROR: [plugin.video.meta] missing translation for choose show
15:00:14 T:15576   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <type 'exceptions.TypeError'>
                                            Error Contents: descriptor 'encode' requires a 'unicode' object but received a 'str'
                                            Traceback (most recent call last):
                                              File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\plugin.video.meta\addon.py", line 177, in <module>
                                                main()
                                              File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\plugin.video.meta\addon.py", line 174, in main
                                                plugin.run()
                                              File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\plugin.video.meta\resources\lib\xbmcswift2\plugin.py", line 333, in run
                                                items = self._dispatch(self.request.path)
                                              File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\plugin.video.meta\resources\lib\xbmcswift2\plugin.py", line 306, in _dispatch
                                                listitems = view_func(**items)
                                              File "C:\Users\Thomas\AppData\Roaming\Kodi\addons\plugin.video.meta\resources\lib\meta\navigation\tvshows.py", line 120, in tv_play_by_name
                                                unicode.encode(s["seriesname"], "ascii", "ignore"), s["year"]) for s in items])
                                            TypeError: descriptor 'encode' requires a 'unicode' object but received a 'str'
                                            -->End of Python script error report<--
```

**Comment by ghost** [comment]
amended my last commit to meta, can you try again? this has turned out to be a bit more complex than i originaly thought :smile:

**Comment by bluezed** [comment]
Yep, seems to be working now :smiley: 

And yes, things like this often turn into bigger projects :wink: 

So... unless you got anything else to add I think it's ok to merge now? Don't think it'll affect any existing installation unless they have your meta-addon installed :+1:

**Comment by ghost** [comment]
once https://github.com/metate/meta4kodi/pull/31 has been merged it should be good to go, otherwise it will cause errors

**Comment by bluezed** [comment]
Great... can you add a note here when it's done please? I'll merge it in then :+1: 

Thanks again for all the effort!

**Comment by ghost** [comment]
yeah, i'll add a note once it's merged

**Comment by ghost** [comment]
merged in meta, but not released yet

**Comment by ghost** [comment]
doing a bit more work on it to make it work when the data isn't present, but it'll require more changes to meta

**Comment by bluezed** [comment]
You still working on this or is it ready for release now?

**Comment by ghost** [comment]
it's ready, but still waiting on metate to do a release so the required changes are available to everyone not running the master branch

**Comment by bluezed** [comment]
One more thing... when the dialogue pops up to ask for "Movie" or "TV Show" it says "Test" at the top. Is that intentional or left over from testing? :wink:

**Comment by ghost** [comment]
woops :smile:
should be fixed now

**Comment by ghost** [comment]
should be good to merge now, meta had another release

**Comment by bluezed** [comment]
@midraal  Just wanted to thank you again for all the work you did for the integration! Much appreciated :+1: 

I've done the release last night and made an announcement on the FTVGuide forum so hopefully people will start using it soon. I certainly have already and works great :smile:

**Pull request merged**